source/extensions/listener_managers/listener_manager/listener_manager_impl.cc

```diff
@@ -238,7 +239,7 @@ Network::SocketSharedPtr ProdListenerComponentFactory::createListenSocket(
           fmt::format("socket type {} not supported for pipes", toString(socket_type)));
     }
     const std::string addr = fmt::format("unix://{}", address->asString());
-    const int fd = server_.hotRestart().duplicateParentListenSocket(addr, worker_index);
+    const int fd = server_.hotRestart().duplicateParentListenSocket(addr, worker_index, listener);
```

Comment:
This listener is the ListenerConfig object not the active TCP/QUIC listener. I'm wondering which field of this object do you want to build the hash map upon, i.e. name or tag? Could you pass in that instead? 

And please keep in mind that the address configured can be Any. When you do the mapping, this need special handling.

Reply:
Oh, this is the "no dynamic cast" problem again *and* I think I misread the chain of inheritance because of two completely different types of things being called listeners, and the class hierarchy in this area being a real spaghetti. :(

I definitely can't pass the keys in instead because the values of the map have to be the objects on which onData can be called. So I need, I guess, the UdpListenerCallbacks, as the base-level type that will do what's needed. (The key would be a tuple of the addr and worker index, which are already the arguments to duplicateParentListenSocket.)

Good point about the Any address, I'll add that to the notes.

Reply:
Turns out no I didn't misread the chain of things, I was thinking of dynamic_casting to `UdpListenerConfig`, which has `listenerWorkerRouter()` on it, which would allow us to deliver packets to the appropriate worker via `deliver()`. So my mistake was that dynamic_casting isn't an option so I'd need to instead make it an `OptRef<UdpListenerConfig>` that gets passed through all these layers.

Alternative option is to capture *either* a UdpListenerConfig *or* an ActiveUdpListener somewhere else in the process and pass it in via a different function, which makes less of a mess (see #29249 for example); but then the question is do we have a preference between:
1. map[addr] -> udpListenerConfig.listenerWorkerRouter(addr).deliver(worker_id, packet)
or
2. map[addr+worker_index] -> ActiveUdpListener.post(packet)

Option 1 transforms addresses to strings in listenerWorkerRouter, and appears to not have a special case for 0.0.0.0 addresses, so maybe option 2 is necessarily the way.

Reply:
Well option 1 turns out to be *much* cleaner now that I'm passing it from a bit further up the stack. The call is made from the main thread (a few calls up it's in a TRY_ASSERT_MAIN_THREAD block), which I *think* is also the thread hot restart mostly runs on. I assume tsan will yell at me at some point in this process if that's incorrect.

Included a comment about 0.0.0.0, implementation for it will be later, and I think can stay out of listenerWorkerRouter (I just have to be sure to pass 0.0.0.0 *into* listenerWorkerRouter if that's the listener the hot restart function chooses.)